**ENSEMBLE MODEL: ENSEMBLING (HARD VOTING)**

**DATA LOADING AND PREPARATION**

In [12]:
import pandas as pd

In [13]:
# Read the CSV
tfidf_df = pd.read_csv("tfidf_sncb.csv", sep='\,', engine='python')

tfidf_df['incident_type'] = tfidf_df['incident_type'].astype('string') 

tfidf_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1011 entries, 0 to 1010
Columns: 828 entries, incident_id to 998
dtypes: float64(826), int64(1), string(1)
memory usage: 6.4 MB


In [14]:
from sklearn.model_selection import train_test_split

# Filter in the Features (the values acquired from the events sequence after TF-IDF)
X = tfidf_df.drop(['incident_type', 'incident_id'], axis=1) 

# Filter in the Target variable (labels / incident types)
y = tfidf_df['incident_type']  

# setting random_state constant to be used in the whole pipeline and guarantee reproducibility
r_state = 123

# Split data into training+validation and testing sets
train_val_X, test_X, train_val_y, test_y = train_test_split(X, 
                                                            y, 
                                                            train_size = 0.8, 
                                                            random_state = r_state, # setting random_state for reproducibility
                                                            stratify = y) # to respect class imbalance in the label column

print(f"The train_val_X pandas df has {len(train_val_X)} rows and {len(train_val_X.columns)} columns.")
print(f"The test_y pandas series has {len(test_y)} rows and 1 column.")

The train_val_X pandas df has 808 rows and 826 columns.
The test_y pandas series has 203 rows and 1 column.


In [15]:
from collections import Counter

# get the size of the smallest incident type class
value_counts = Counter(train_val_y)
min_class_setsize = min(value_counts.values())

print(f"In RepeatedStratifiedKFold() function, the parameter n_splits has to be set atmost to {min_class_setsize}, due to class imbalance in the label column.")

In RepeatedStratifiedKFold() function, the parameter n_splits has to be set atmost to 3, due to class imbalance in the label column.


**WEIGHTING FOR ENSEMBLE CLASSIFIER (HARD VOTING)**

In [16]:
import numpy as np

f1score_rf = pd.read_csv("weighted_f1_rf.csv", sep='\,', engine='python')
f1score_nn = pd.read_csv("weighted_f1_nn.csv", sep='\,', engine='python')
f1score_gb = pd.read_csv("weighted_f1_gb.csv", sep='\,', engine='python')

weight_rf = np.mean(f1score_rf['score'])
weight_nn = np.mean(f1score_nn['score'])
weight_gb = np.mean(f1score_gb['score'])
weight_sum = weight_rf + weight_nn + weight_gb

# raw weights
raw_weights = [weight_rf, weight_nn, weight_gb]

# normalized weights
norm_weights = raw_weights / weight_sum

print(f"Original Mean Weighted F1-Scores: {raw_weights}")
print(f"Normalized Weights:               {norm_weights}")

Original Mean Weighted F1-Scores: [np.float64(0.5739477552752065), np.float64(0.5808108262467909), np.float64(0.6389097766681082)]
Normalized Weights:               [0.31998544 0.32381171 0.35620285]


**MODEL TRAINING AND VALIDATION**

In [17]:
from sklearn.model_selection import RepeatedStratifiedKFold
# from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, StackingClassifier, VotingClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, VotingClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, f1_score
from datetime import datetime

# Define array of base models
base_models = [
    ('Random Forest', RandomForestClassifier(random_state = r_state)),
    ('Neural Networks', MLPClassifier(random_state = r_state)),
    ('Gradient Boosting', GradientBoostingClassifier(random_state = r_state))
]

# Define Ensembling (Hard Voting)
model_ensembling = VotingClassifier(estimators = base_models, 
                                voting = 'hard', 
                                weights = norm_weights)

# Set up cross-validation
rskf = RepeatedStratifiedKFold(n_splits = min_class_setsize, 
                               n_repeats = 34, 
                               random_state = r_state)

# Start timing
start_time = datetime.now()

# Initialize lists to store metrics
accuracy_scores = []
weighted_f1_scores = []
micro_f1_scores = []
macro_f1_scores = []
fold = 1

# Cross-validation loop
for train_idx, val_idx in rskf.split(train_val_X, train_val_y):

    # Uncomment the print statement for debugging only
    #print(f"Starting model training and validation on fold {fold}")
    
    # Split data
    fold_train_X, fold_val_X = train_val_X.iloc[train_idx], train_val_X.iloc[val_idx]
    fold_train_y, fold_val_y = train_val_y.iloc[train_idx], train_val_y.iloc[val_idx]

    # Train the ensemble model
    model_ensembling.fit(fold_train_X, fold_train_y)  # Train the ensemble model
    
    # Predict on the test set
    fold_pred_y = model_ensembling.predict(fold_val_X)
    
    # Compute metrics
    accuracy_scores.append(accuracy_score(fold_val_y, fold_pred_y))
    weighted_f1_scores.append(f1_score(fold_val_y, fold_pred_y, average='weighted'))
    micro_f1_scores.append(f1_score(fold_val_y, fold_pred_y, average='micro'))
    macro_f1_scores.append(f1_score(fold_val_y, fold_pred_y, average='macro'))
    
    fold += 1

# End timing
end_time = datetime.now()
elapsed_time = end_time - start_time

# Aggregate results
print(f"The training and validation of model_ensembling took {elapsed_time.total_seconds():.2f} seconds across {rskf.get_n_splits()} iterations.\n")

print(f"Mean Accuracy:          {np.mean(accuracy_scores):.8f} ± {np.std(accuracy_scores):.8f}")
print(f"Mean Weighted F1-Score: {np.mean(weighted_f1_scores):.8f} ± {np.std(weighted_f1_scores):.8f}")
print(f"Mean Micro F1-Score:    {np.mean(micro_f1_scores):.8f} ± {np.std(micro_f1_scores):.8f}")
print(f"Mean Macro F1-Score:    {np.mean(macro_f1_scores):.8f} ± {np.std(macro_f1_scores):.8f}\n")

print(f"Each fold had {len(fold_train_X)} entries for training and {len(fold_val_X)} for validation.")

/usr/local/Caskroom/miniconda/base/envs/ulbdm/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/Caskroom/miniconda/base/envs/ulbdm/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/Caskroom/miniconda/base/envs/ulbdm/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/Caskroom/miniconda/base/envs/ulbdm/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and 

The training and validation of model_rf took 1733.67 seconds across 102 iterations.

Mean Accuracy:          0.65358006 ± 0.02215949
Mean Weighted F1-Score: 0.63647744 ± 0.02239695
Mean Micro F1-Score:    0.65358006 ± 0.02215949
Mean Macro F1-Score:    0.34742773 ± 0.02612907

Each fold had 539 entries for training and 269 for validation.


**The training and validation of model_ensembling took 1733.67 seconds across 102 iterations.**

Mean Accuracy:          0.65358006 ± 0.02215949

Mean Weighted F1-Score: 0.63647744 ± 0.02239695

Mean Micro F1-Score:    0.65358006 ± 0.02215949

Mean Macro F1-Score:    0.34742773 ± 0.02612907

**Each fold had 539 entries for training and 269 for validation.**

**TEST SET RESULTS**

In [18]:
from sklearn.metrics import classification_report, confusion_matrix

# Train the ensemble model
model_ensembling.fit(train_val_X, train_val_y)

# Predict on the test set
test_pred_y = model_ensembling.predict(test_X)

# Compute and display metrics
print(f"The model classified correctly {sum(test_y == test_pred_y)} entries from a total of {len(test_X)}.\n")

print(f"Accuracy on test set:          {accuracy_score(test_y, test_pred_y)}")
print(f"Weighted F1-Score on test set: {f1_score(test_y, test_pred_y, average='weighted')}\n")

print("F1-Score per class\n")

# Generate classification report
report = classification_report(test_y, test_pred_y, output_dict=True, zero_division=0)

# Display F1-score per class
for class_label, metrics in report.items():
    if isinstance(metrics, dict) and 'f1-score' in metrics:
        print(f"Class {class_label}: F1-Score = {metrics['f1-score']:.6f}")

print("\nAccuracy per class\n")

# Display F1-score per class
for class_label, metrics in report.items():
    if isinstance(metrics, dict) and 'recall' in metrics:
        print(f"Class {class_label}: Recall = {metrics['recall']:.6f}") # Recall is equivalent to per-class accuracy

/usr/local/Caskroom/miniconda/base/envs/ulbdm/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


The model classified correctly 132 entries from a total of 203.

Accuracy on test set:          0.6502463054187192
Weighted F1-Score on test set: 0.6311644591989505

F1-Score per class

Class 11: F1-Score = 0.333333
Class 13: F1-Score = 0.789116
Class 14: F1-Score = 0.641509
Class 16: F1-Score = 0.000000
Class 17: F1-Score = 0.000000
Class 2: F1-Score = 0.807692
Class 3: F1-Score = 0.000000
Class 4: F1-Score = 0.571429
Class 6: F1-Score = 0.000000
Class 7: F1-Score = 0.000000
Class 9: F1-Score = 0.600000
Class 99: F1-Score = 0.410959
Class macro avg: F1-Score = 0.346170
Class weighted avg: F1-Score = 0.631164

Accuracy per class

Class 11: Recall = 0.200000
Class 13: Recall = 0.906250
Class 14: Recall = 0.566667
Class 16: Recall = 0.000000
Class 17: Recall = 0.000000
Class 2: Recall = 0.875000
Class 3: Recall = 0.000000
Class 4: Recall = 0.500000
Class 6: Recall = 0.000000
Class 7: Recall = 0.000000
Class 9: Recall = 0.521739
Class 99: Recall = 0.428571
Class macro avg: Recall = 0.3331

**The model classified correctly 132 entries from a total of 203.**

Accuracy on test set:          0.6502463054187192

Weighted F1-Score on test set: 0.6311644591989505

**F1-Score per class**

Class 11: F1-Score = 0.333333

Class 13: F1-Score = 0.789116

Class 14: F1-Score = 0.641509

Class 16: F1-Score = 0.000000

Class 17: F1-Score = 0.000000

Class 2: F1-Score = 0.807692

Class 3: F1-Score = 0.000000

Class 4: F1-Score = 0.571429

Class 6: F1-Score = 0.000000

Class 7: F1-Score = 0.000000

Class 9: F1-Score = 0.600000

Class 99: F1-Score = 0.410959

Class macro avg: F1-Score = 0.346170

Class weighted avg: F1-Score = 0.631164

**SAVE AND EXPORT RESULTS**

In [19]:
"""
# Create DataFrame for Scores
accuracy_ensembling_df = pd.DataFrame({'score': accuracy_scores})
weighted_f1_ensembling_df = pd.DataFrame({'score': weighted_f1_scores})
micro_f1_ensembling_df = pd.DataFrame({'score': micro_f1_scores})
macro_f1_ensembling_df = pd.DataFrame({'score': macro_f1_scores})

# Export the DataFrame to a CSV file
accuracy_ensembling_df .to_csv('accuracy_ensembling.csv', index=False)
weighted_f1_ensembling_df.to_csv('weighted_f1_ensembling.csv', index=False)
micro_f1_ensembling_df.to_csv('micro_f1_ensembling.csv', index=False)
macro_f1_ensembling_df.to_csv('macro_f1_ensembling.csv', index=False)
"""